src/datastore/postgres-notifier.ts

```diff
@@ -99,15 +109,27 @@ export class PgNotifier {
   }
 
   public async close() {
-    await this.subscriber.unlisten(this.pgChannelName);
-    await this.subscriber.close();
+    await this.subscriber?.unlisten(this.pgChannelName);
+    await this.subscriber?.close();
   }
 
   private async notify(notification: PgNotification) {
-    await this.subscriber
-      .notify(this.pgChannelName, { notification: notification })
-      .catch(error =>
-        logError(`Error sending PgNotifier notification of type: ${notification.type}`, error)
-      );
+    if (this.subscriber) {
+      await this.subscriber
+        .notify(this.pgChannelName, { notification: notification })
+        .catch(async error => {
+          if (error instanceof Error && error.message.includes('not queryable')) {
+            // The postgres client has lost the connection or has otherwise become corrupt. We will attempt to
+            // reconnect so we can continue sending events.
+            // There's no point in calling `unlisten` before closing the current client since it requires a query.
+            await this.subscriber?.close();
+            await this.doConnect();
```

Comment:
Does `doConnect` have retry logic? If so, it seems like we could hit this condition, then try to connect while the db is still unavailable, and end up in another error state?

Reply:
That's a good question. It doesn't have logic in this file but probably in the library. I'll take a look.

Reply:
It does try to reconnect on error and it retries an infinite number of times until it succeeds, so I think we're good: https://github.com/andywer/pg-listen/blob/master/src/index.ts#L93-L125

However, I can see a scenario where we lose some notifications that would come while a reconnection is still pending (the code as it is right now would only retry the original notification that triggered the reconnect).

I could add some sort of queue that would retry any notification on failure up to a fixed number of times, but that may be outside the scope of this PR. Thoughts?

Reply:
The dropped events seem okay to me in this scenario. Were you able to reproduce this behavior locally? Right now this error state results in the API still working (without websockets), we'd just want to ensure a possible regression around here didn't make things worse.

Reply:
I was just able to reproduce this locally, it came down to restarting the database just at the right time when a notification was about to be sent.

As it turns out, the library **does** reconnect when this error happens but it still logs the error. Adding this `close()`/`doConnect()` combination was revealed to be useless because `close()` actually **destroys** the whole error callback without finishing it 🤣 so `doConnect()` is never really called.

@CharlieC3 when you saw these errors on #960, did they continue to appear constantly from then on? Or did they just appear while the database was being restarted? My guess is that the series of errors you were seeing were from all the different message subscriptions that tried to notify at the same time or very close to each other, so this may actually be working as intended.